From 25b8aaf464609b1f77eef4d64af36b25c99c1599 Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Thu, 28 Jun 2018 19:24:06 -0700
Subject: [PATCH] MANTA-3798 waferlock should only watch postgres/moray within
 the same shard MANTA-3799 waferlock should avoid short cueball decoherence
 with ZK Reviewed by: Cody Peter Mello <cody.mello@joyent.com> Approved by:
 Cody Peter Mello <cody.mello@joyent.com>

---
 lib/sapi.js                       | 9 +++++++++
 package.json                      | 2 +-
 sapi_manifests/waferlock/template | 3 ++-
 server.js                         | 2 ++
 4 files changed, 14 insertions(+), 2 deletions(-)

diff --git a/lib/sapi.js b/lib/sapi.js
index b3e254a..967579e 100644
--- a/lib/sapi.js
+++ b/lib/sapi.js
@@ -31,6 +31,8 @@ function SapiPoller(options) {
 	mod_assert.number(options.minPoll, 'options.minPoll');
 	mod_assert.number(options.maxPoll, 'options.maxPoll');
 
+	mod_assert.optionalString(options.shard, 'options.shard');
+
 	this.sp_log = options.log.child({
 		component: 'SapiPoller',
 		service: options.service
@@ -38,6 +40,7 @@ function SapiPoller(options) {
 	this.sp_pool = options.pool;
 	this.sp_url = options.url;
 	this.sp_suffix = options.dns_domain;
+	this.sp_shard = options.shard;
 
 	var parts = options.service.split('/');
 	mod_assert.equal(parts.length, 2);
@@ -179,6 +182,12 @@ SapiPoller.prototype.state_insts = function (S) {
 		added.forEach(function (id) {
 			var inst = newids[id];
 			var dc = inst.metadata['DATACENTER'];
+			var shard = inst.metadata['SHARD'];
+			if (shard && self.sp_shard && shard !== self.sp_shard) {
+				self.sp_log.trace('dropping sapi inst %s: ' +
+				    'shard does not match', id);
+				return;
+			}
 			if (dc === undefined)
 				dc = self.sp_dc;
 			if (dc === undefined) {
diff --git a/package.json b/package.json
index 2a6e37a..f7637b8 100644
--- a/package.json
+++ b/package.json
@@ -29,7 +29,7 @@
     "restify-clients": "^2.2.0",
     "vasync": "^2.2.0",
     "verror": "^1.10.0",
-    "zkstream": "^0.10.1"
+    "zkstream": "^0.11.2"
   },
   "devDependencies": {
     "eslint": "4.19.1",
diff --git a/sapi_manifests/waferlock/template b/sapi_manifests/waferlock/template
index 7609bf9..73334da 100644
--- a/sapi_manifests/waferlock/template
+++ b/sapi_manifests/waferlock/template
@@ -37,7 +37,7 @@
 	{{^dns_domain}}
 		"paths": [],
 		"domains": [
-			"moray.{{{DOMAIN_NAME}}}",
+			"{{{SHARD}}}.moray.{{{DOMAIN_NAME}}}",
 			"ops.{{{DOMAIN_NAME}}}",
 			"madtom.{{{DOMAIN_NAME}}}",
 			"pgstatsmon.{{{DOMAIN_NAME}}}",
@@ -46,6 +46,7 @@
 		],
 		"sapi_url": "{{{SAPI_URL}}}",
 		"sapi_services": ["manta/moray"],
+		"shard": "{{{SHARD}}}",
 		"dns_domain": "{{{DNS_DOMAIN}}}"
 	{{/dns_domain}}
 }
diff --git a/server.js b/server.js
index eafb9db..fa6336a 100644
--- a/server.js
+++ b/server.js
@@ -48,6 +48,7 @@ mod_assert.number(config.sapiPollingInterval.min,
     'config.sapiPollingInterval.min');
 mod_assert.number(config.sapiPollingInterval.max,
     'config.sapiPollingInterval.max');
+mod_assert.optionalString(config.shard, 'config.shard');
 
 mod_assert.optionalArrayOfString(config.paths, 'config.paths');
 mod_assert.optionalArrayOfString(config.domains, 'config.domains');
@@ -195,6 +196,7 @@ AppFSM.prototype.state_setupSapi = function (S) {
 			dns_domain: config.dns_domain,
 			minPoll: config.sapiPollingInterval.min,
 			maxPoll: config.sapiPollingInterval.max,
+			shard: config.shard,
 			agent: self.af_agent
 		});
 		/*
-- 
2.21.0

